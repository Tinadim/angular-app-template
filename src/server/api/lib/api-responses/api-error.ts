import { serverConfig } from '@server/config/server-config';

export class ApiError extends Error {
    readonly code: ErrorCode;
    readonly message: string;
    readonly errors?: ErrorDetail[];

    constructor(code: ErrorCode, message: string, errors?: ErrorDetail[]) {
        super(message);
        this.code = code;
        this.message = message;
        this.errors = errors;
    }

    toJson() {
        const errorObject: { [key: string]: any } = {
            code: this.code,
            message: this.message,
            errors: this.errors
        };

        if (serverConfig.isProduction) {
            errorObject.stack = this.stack;
        }

        return errorObject;
    }
}

export interface ErrorDetail {
    field?: string;
    code: string;
    message: string;
}

export const enum ErrorMessages {
    Forbidden = 'You don\'t have access to the requested resource',
    InvalidParameters = 'Invalid parameters provided',
    NotFound = 'Resource not found',
    Unauthorized = 'Authentication required',
    UnexpectedError = 'Unexpected error',
}

export enum ErrorCode {
    EntityNotFound = 'entityNotFound',
    Forbidden = 'accessDenied',
    InvalidParameters = 'invalidParameters',
    MissingRequiredParameter = 'missingRequiredParameter',
    Unauthorized = 'authenticationRequired',
    UnexpectedError= 'unexpectedError',
}

