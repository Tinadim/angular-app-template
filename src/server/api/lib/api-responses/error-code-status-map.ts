import { ErrorCode } from '@server/lib/exceptions/error-code';

const map: { [key in ErrorCode]: number } = {
    [ErrorCode.EntityNotFound]: 404,
    [ErrorCode.Forbidden]: 403,
    [ErrorCode.InvalidParameters]: 400,
    [ErrorCode.MissingRequiredParameter]: 400,
    [ErrorCode.Unauthorized]: 401,
    [ErrorCode.UnexpectedError]: 500,
};

export function getStatusForErrorCode(errorCode: ErrorCode) {
    return map[errorCode] || 500;
}
