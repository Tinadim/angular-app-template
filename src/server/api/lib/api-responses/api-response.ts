import * as HttpStatusCodes from 'http-status-codes';
import { Response } from 'express';
import { ApplicationError } from '@server/lib/exceptions/application-error';
import { getStatusForErrorCode } from '@server/api/lib/api-responses/error-code-status-map';

export class ApiResponse {
    readonly status: number;
    readonly success: boolean;
    readonly data?: object;
    readonly error?: ApplicationError;

    constructor(status: number, data?: object, error?: ApplicationError) {
        this.status = status;
        this.data = data;
        this.error = error;
        this.success = this.isSuccessResponse();
    }

    static Ok(data?: object) {
        return new ApiResponse(HttpStatusCodes.OK, data);
    }

    static Created(data?: object) {
        return new ApiResponse(HttpStatusCodes.CREATED, data);
    }

    static NoContent() {
        return new ApiResponse(HttpStatusCodes.NO_CONTENT);
    }

    static TooManyRequests() {
        return new ApiResponse(HttpStatusCodes.TOO_MANY_REQUESTS);
    }

    static fromApplicationError(error: ApplicationError) {
        const status = getStatusForErrorCode(error.code);
        return new ApiResponse(status, undefined, error);
    }

    toJson() {
        return {
            data: this.data,
            error: this.error && this.error.toJson(),
        };
    }

    send(res: Response) {
        res.status(this.status).json(this.toJson());
    }

    private isSuccessResponse() {
        return this.status < 400;
    }
}
