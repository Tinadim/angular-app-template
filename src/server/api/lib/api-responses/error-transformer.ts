import ajv, { ErrorObject } from 'ajv';
import { get, map } from 'lodash';
import { ApiError, ErrorCode, ErrorDetail, ErrorMessages } from './api-error';
import { ApiResponse } from './api-response';

const keywordMap: { [key: string]: ErrorCode | undefined } = {
    required: ErrorCode.MissingRequiredParameter,
};

function mapAjvError(ajvError: ErrorObject): ErrorDetail {
    let { message } = ajvError;
    const code = keywordMap[ajvError.keyword] || ErrorCode.InvalidParameters;
    const params = get(ajvError, 'params', {}) as any;
    const fieldName = (ajvError.dataPath || params.property || params.missingProperty ||
        params.additionalProperties || '').replace('.', '');
    if (ajvError.keyword === 'enum') {
        const allowedValues = get(ajvError, 'params.allowedValues');
        message += `: ${allowedValues}`;
    }

    return {
        code,
        message,
        field: fieldName,
    };
}

export function fromAjvError(ajvError: ajv.ValidationError): ApiResponse {
    const apiError = new ApiError(
        ErrorCode.InvalidParameters,
        ErrorMessages.InvalidParameters,
        map(ajvError.errors, mapAjvError)
    );
    return ApiResponse.BadRequest(apiError);
}
