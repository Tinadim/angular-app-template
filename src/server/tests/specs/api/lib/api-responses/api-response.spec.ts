import { ApiResponse } from '@server/api/lib/api-responses/api-response';

describe('api-response.spec.ts', () => {
    function testApiResponse(response: ApiResponse) {
        const result = response.toJson();
        expect(result).toMatchSnapshot();
    }
    it('Should create a 200 success response', () => {
        const response = ApiResponse.Ok();
        testApiResponse(response);
    });

    it('Should create a 200 success response with the provided data', () => {
        const response = ApiResponse.Ok({ foo: 'bar' });
        testApiResponse(response);
    });

    it('Should create a 201 created response', () => {
        const response = ApiResponse.Created();
        testApiResponse(response);
    });

    it('Should create a 201 created response with the provided data', () => {
        const response = ApiResponse.Created({ foo: 'bar' });
        testApiResponse(response);
    });

    it('Should create a 204 no-content response', () => {
        const response = ApiResponse.NoContent();
        testApiResponse(response);
    });

    it('Should create a 429 too many requests response', () => {
        const response = ApiResponse.TooManyRequests();
        testApiResponse(response);
    });
});
