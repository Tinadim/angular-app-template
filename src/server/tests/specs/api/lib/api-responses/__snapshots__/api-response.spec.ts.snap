// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-response.spec.ts Should create a 200 success response 1`] = `
Object {
  "data": undefined,
  "error": undefined,
}
`;

exports[`api-response.spec.ts Should create a 200 success response with the provided data 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "error": undefined,
}
`;

exports[`api-response.spec.ts Should create a 201 created response 1`] = `
Object {
  "data": undefined,
  "error": undefined,
}
`;

exports[`api-response.spec.ts Should create a 201 created response with the provided data 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "error": undefined,
}
`;

exports[`api-response.spec.ts Should create a 204 no-content response 1`] = `
Object {
  "data": undefined,
  "error": undefined,
}
`;

exports[`api-response.spec.ts Should create a 429 too many requests response 1`] = `
Object {
  "data": undefined,
  "error": undefined,
}
`;
